crates/ironrdp-pdu/src/geometry.rs

```diff
@@ -3,34 +3,118 @@ use std::io;
 
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
 
-/// An **inclusive** rectangle.
-///
-/// This struct is defined as an **inclusive** rectangle.
-/// That is, the pixel at coordinate (right, bottom) is included in the rectangle.
+pub trait RectangleImpl {
+    const WIDTH_VALIDATION_ERROR: &'static str;
+    const HEIGHT_VALIDATION_ERROR: &'static str;
+
+    fn calculate_width(left: u16, right: u16) -> u16;
+    fn calculate_height(top: u16, bottom: u16) -> u16;
+
+    fn is_valid_width(left: u16, right: u16) -> bool;
+    fn is_valid_height(top: u16, bottom: u16) -> bool;
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Rectangle {
+pub struct BaseRectangle<I: RectangleImpl> {
     pub left: u16,
     pub top: u16,
     pub right: u16,
     pub bottom: u16,
+    _phantom_data: std::marker::PhantomData<I>,
 }
```

Comment:
Instead of having a "BaseRectangle" with a PhantomData field, I feel we should have two structs, `InclusiveRectangle` and `ExclusiveRectangle`, both implementing the `Rectangle` trait so we can still define operations that can be applied to both in a generic way as necessity arise.

The `Rectangle` would defines some of the operations currently defined on `BaseRectangle` (`height`, `width`, …).
Other methods such as `union` may be either free standing functions taking generic `Rectangle` parameters (e.g.: `union(r1, r2)`), or defined on the `Rectangle` trait with a default implementation, or better, on a `RectangleExt` trait automatically implemented for any `Rectangle`.

Reply:
Yep, I agree, I did not like that change too, I'll revisit that

Reply:
@CBenoit I updated the approach to `Rectangle`, please take a look -- what do you think about it?

Reply:
I actually like it better! I’ll resolve this comment and add new ones for simplicity